[EuiDataGrid] Add `onChange` callbacks for display selector changes

constancecchen opened the pull request:
### Summary

Per feedback from Caroline and Vlad, we should expose internal `gridStyle` and `rowHeightsOptions` state in an `onChange` callback:

> In the future, we should provide a grid example that manually adjusts its content based on the selected display size. Does that display selector have a callback function that consumers can use?
(From Caroline)

> I had some thought about height switcher, what we can do if devs will need current state of rowHeightsOptions to store it in something like saved object.
(From Vlad)

After discussion with @chandlerprall, the API we settled on was adding these callbacks in `gridStyle.onChange` and `rowHeightsOptions.onChange`.

### Screencaps

<img width="1103" alt="" src="https://user-images.githubusercontent.com/549407/144127770-0cbea5b0-d264-40e6-aa51-75a87e47dbb8.png">

https://user-images.githubusercontent.com/549407/144127724-c0dba6d8-6329-4a9a-b36b-15e67b931e4e.mp4

### Checklist

~- [ ] Check against **all themes** for compatibility in both light and dark modes~
~- [ ] Checked in **mobile**~
~- [ ] Checked in **Chrome**, **Safari**, **Edge**, and **Firefox**~

- [x] Props have proper **autodocs** and **[playground toggles](https://github.com/elastic/eui/blob/main/wiki/documentation-guidelines.md#adding-playground-toggles)**
- [x] Added **[documentation](https://github.com/elastic/eui/blob/main/wiki/documentation-guidelines.md)**

~- [ ] Checked **[Code Sandbox](https://codesandbox.io/)** works for any docs examples~

- [x] Added or updated **[jest](https://github.com/elastic/eui/blob/main/wiki/testing.md) and [cypress](https://github.com/elastic/eui/blob/main/wiki/cypress-testing.md) tests**

~- [ ] Checked for **breaking changes** and labeled appropriately~
~- [ ] Checked for **accessibility** including keyboard-only and screenreader modes~

- [x] A **[changelog](https://github.com/elastic/eui/blob/main/wiki/documentation-guidelines.md#changelog)** entry exists and is marked appropriately

constancecchen commented on src/components/datagrid/controls/display_selector.tsx:
```diff
@@ -110,10 +124,14 @@ export const useDataGridDisplaySelector = (
+    (density: string) => {
+      _setGridDensity(density);
+      setUserGridStyles(densityStyles[density]);
+      onDensityChange?.({ ...initialStyles, ...densityStyles[density] });
```
@chandlerprall I went back and forth on this a bit and I'm still not 100% - should we send back the entirety of the current `gridStyles` / `rowHeightsOptions` currently in state, or should we only send back the recent changes (i.e., without `...initialStyles`?)

chandlerprall replied:
I think it makes sense to send back the entirety of the current configuration, two considerations I can think of:

* in the future, if we allow users to change other display choices it will automatically be accounted for in the callback, and applications are already used to receiving that data
* if an application exposes their own controls over e.g. font size, returning the whole config should help that application's data handling of the changes

We should remove the `onChange` entry from the callback's values so it only represents the configuration values, however

chandlerprall commented on src/components/datagrid/controls/display_selector.tsx:
```diff
@@ -137,8 +155,12 @@ export const useDataGridDisplaySelector = (
       setRowHeightSelection(option);
       setUserRowHeightsOptions(rowHeightsOptions);
+      onRowHeightChange?.({
+        ...initialRowHeightsOptions,
```
We should remove the onChange entry from the callback's values so it only represents the configuration values

constancecchen replied:
Whoops, awesome catch!

constancecchen replied:
https://github.com/elastic/eui/pull/5424/commits/fb5bb6326c3425d5d702967f5ef1320fd31f43ea#diff-4955b4980313a3447c772a7d0d87d126f3c4db9cd066496d59e0b27ac80e0390

constancecchen commented on src/components/datagrid/controls/display_selector.tsx:
```diff
@@ -137,8 +155,12 @@ export const useDataGridDisplaySelector = (
 
       setRowHeightSelection(option);
       setUserRowHeightsOptions(rowHeightsOptions);
+      onRowHeightChange?.({
```
Super good call! I also realized I was forgetting about lineCount changes while working on #5428 🙈

constancecchen replied:
Speaking of #5428 I think I may end up putting this back in draft mode until that PR lands and then adjusting this PR accordingly. Why I say that is because I may end up tweaking this PR to call `onRowHeightChange` on `useEffect` whenever `userRowHeightsOptions`/`userGridStyles` changes rather than manually within callbacks. That feels more "React-y" and also DRYs out the scenario where lineCount changes.

constancecchen replied:
https://github.com/elastic/eui/pull/5424/commits/fb5bb6326c3425d5d702967f5ef1320fd31f43ea

constancecchen commented:
OK, I updated this PR to use a new `onUpdateEffect` helper (fires on update only, not mount). I ended up force pushing because the code was so different, sorry about that - I recommend re-following along by commit. Quick TL;DR of changes:

- Callbacks are now invoked in an effect and not manually in each state update, which is much more DRY and feels much more React-y
- Removed passing the `onChange` function back into the callbacks (inception BWAAM)
- I removed the misc HTML cleanup commit since I opened up a separate PR for that (#5429)
- I removed the extra `callbacks` arg I was passing into `useDataGridDisplaySelector` - I have no idea why I thought that was necessary when I was already passing in `gridStyles` and `rowHeightsOptions` 🤦‍♀️ 
- Documentation example remained functionally the same with some minor cleanup - I stopped saving the returned objs in state since that didn't work well when combined with #5428's reset button, and added a console.log for easier debugging

Overall this feels much cleaner than before! I also tested this rebased against #5248 and the onChange callbacks are correctly invoked when the reset button is pressed without any extra code 🎉 The merges will have very minor conflicts as well, so either PR can land first without significant issue

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5424/

miukimiu reviewed (approved):
Thanks, @constancecchen. It looks great! 

I tested the example code and it works well. The reset button from #5428 now makes even more sense. 👍🏽 

Just one small detail. Is it possible to improve the example to also preserve the display settings when the page changes?

https://user-images.githubusercontent.com/2750668/144432974-b20353d0-1870-470b-b5b2-fe94bbac25cc.mov

constancecchen commented:
Oh super interesting, thanks Elizabet, I hadn't tried navigating away. Yes, it's possible, I need to change where I instantiate the `localStorage.getItem` call into the component itself. I'll do that here shortly.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5424/

chandlerprall reviewed (approved):
Code changes LGTM, tests are great! Tested locally & in the preview docs.

constancecchen commented:
Thanks y'all!
